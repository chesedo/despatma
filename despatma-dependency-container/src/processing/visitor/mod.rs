use despatma_visitor::visitor_mut;

use super::{ChildDependency, Container, Dependency};

pub use add_wildcard_lifetime::AddWildcardLifetime;
pub use extract_async::ExtractAsync;
pub use extract_box_type::ExtractBoxType;
pub use extract_lifetime::ExtractLifetime;
pub use impl_trait_but_registered_concrete::ImplTraitButRegisteredConcrete;
pub use impl_trait_fields::ImplTraitFields;
pub use link_dependencies::LinkDependencies;
pub use owning_managed_dependency::OwningManagedDependency;
pub use replace_impl_generics_with_concrete::ReplaceImplGenericsWithConcrete;
pub use unsupported_registered_types::UnsupportedRegisteredTypes;
pub use wrap_box_type::WrapBoxType;

mod add_wildcard_lifetime;
mod extract_async;
mod extract_box_type;
mod extract_lifetime;
mod impl_trait_but_registered_concrete;
mod impl_trait_fields;
mod link_dependencies;
mod owning_managed_dependency;
mod replace_impl_generics_with_concrete;
mod unsupported_registered_types;
mod wrap_box_type;

visitor_mut!(
    #[helper_tmpl = {
        for dependency in container.dependencies.iter() {
            visitor.visit_dependency_mut(&mut dependency.borrow_mut());
        }
    }]
    Container,
    #[helper_tmpl = {
        for dependency in dependency.dependencies.iter_mut() {
            visitor.visit_child_dependency_mut(dependency);
        }
    }]
    Dependency,
    #[helper_tmpl = {
        visitor.visit_dependency_mut(&mut child_dependency.inner.borrow_mut());
    }]
    ChildDependency,
);

/// A visitor used for validation
/// If the visitor found any errors then they should be emit in [emit_errors].
pub trait ErrorVisitorMut: VisitorMut + Sized {
    fn new() -> Self;

    fn emit_errors(self) {}
}
