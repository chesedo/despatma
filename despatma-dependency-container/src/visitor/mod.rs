use crate::container::{ChildDependency, Container, Dependency};

mod check_wiring;
mod fix_async_tree;
mod impl_trait_but_registered_concrete;
mod set_output_spans;

pub use check_wiring::CheckWiring;
use despatma_visitor::{visitor, visitor_mut};
pub use fix_async_tree::FixAsyncTree;
pub use impl_trait_but_registered_concrete::ImplTraitButRegisteredConcrete;
pub use set_output_spans::SetOutputSpans;

visitor!(
    #[helper_tmpl = {
        for dependency in container.dependencies.values() {
            visitor.visit_dependency(&dependency.borrow());
        }
    }]
    Container,
    #[helper_tmpl = {
        for child_dependency in &dependency.dependencies {
            visitor.visit_child_dependency(child_dependency);
        }
    }]
    Dependency,
    ChildDependency,
);

/// A visitor used to validate the struct that will be turned into a dependency container.
/// If the visitor found any errors then they should be emit in [emit_errors].
pub trait ErrorVisitor: Visitor {
    fn emit_errors(self);
}

// A mutable visitor used to update any dependencies or their children
visitor_mut!(
    #[helper_tmpl = {
        for dependency in container.dependencies.values_mut() {
            visitor.visit_dependency_mut(&mut dependency.borrow_mut());
        }
    }]
    Container,
    #[helper_tmpl = {
        for child_dependency in &mut dependency.dependencies {
            visitor.visit_child_dependency_mut(child_dependency);
        }
    }]
    Dependency,
    ChildDependency,
);
