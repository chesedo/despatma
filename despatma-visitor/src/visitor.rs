use convert_case::{Case, Casing};
use despatma_lib::{AnnotatedType, KeyValue, SimpleType};
use proc_macro2::{Span, TokenStream, TokenTree};
use quote::{format_ident, quote};
use syn::parse::{Parse, ParseStream, Result};
use syn::punctuated::Punctuated;
use syn::{Ident, Token};

/// Model for holding the input passed to the visitor macro
/// It expects a stream in the following format:
/// ```text
/// ConcreteType,
///
/// dyn DynamicType,
///
/// #[no_defuault]
/// NoDefault,
///
/// #[helper_tmpl = {visitor.visit_button(window.button);}]
/// CustomTemplate,
/// ```
///
/// Thus, it takes a list of types that will be visited.
/// A type can be concrete or dynamic.
///
/// Options can also be passed to the type:
/// - `no_default` to turn-off the defualt implementation for the trait method.
/// - 'helper_tmpl` to be filled into the helper template for traversing a types internal structure.
#[cfg_attr(test, derive(Eq, PartialEq, Debug))]
pub struct VisitorFunction {
    types: Punctuated<AnnotatedType<SimpleType>, Token![,]>,
}

/// Make VisitorFunction parsable
impl Parse for VisitorFunction {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(VisitorFunction {
            types: input.parse_terminated(AnnotatedType::parse, Token![,])?,
        })
    }
}

impl VisitorFunction {
    /// Expand the visitor model into its implementation
    pub fn expand(&self) -> TokenStream {
        // Store each of the three parts
        let mut trait_functions: Vec<TokenStream> = Vec::new();
        let mut helpers: Vec<TokenStream> = Vec::new();
        let mut visitables: Vec<TokenStream> = Vec::new();

        // Loop over each type given
        for t in self.types.iter() {
            let elem_name = Ident::new(
                &t.inner_type.ident.to_string().to_case(Case::Snake),
                t.inner_type.ident.span(),
            );
            let elem_type = &t.inner_type;
            let fn_name = format_ident!("visit_{}", elem_name);
            let options = Options::new(&t.attrs.options);

            // Get trait function
            if options.no_default {
                trait_functions.push(quote! {
                    fn #fn_name(&mut self, #elem_name: &#elem_type);
                })
            } else {
                trait_functions.push(quote! {
                    fn #fn_name(&mut self, #elem_name: &#elem_type) {
                        #fn_name(self, #elem_name)
                    }
                })
            };

            // Get helper function
            if options.has_helper {
                if let Some(inner) = options.helper_tmpl {
                    helpers.push(quote! {
                        pub fn #fn_name<V>(visitor: &mut V, #elem_name: &#elem_type)
                        where
                            V: Visitor + ?Sized,
                        {
                            #inner
                        }
                    });
                } else {
                    let unused_elem_name = format_ident!("_{}", elem_name);
                    helpers.push(quote! {
                        pub fn #fn_name<V>(_visitor: &mut V, #unused_elem_name: &#elem_type)
                        where
                            V: Visitor + ?Sized,
                        {
                        }
                    });
                }
            };

            // Make visitable
            visitables.push(quote! {
                impl Visitable for #elem_type {
                    fn apply(&self, visitor: &mut impl Visitor) {
                        visitor.#fn_name(self);
                    }
                }
            });
        }

        // Built complete visitor implementation
        quote! {
            pub trait Visitor {
                #(#trait_functions)*
            }

            #(#helpers)*

            trait Visitable {
                fn apply(&self, visitor: &mut impl Visitor);
            }
            #(#visitables)*
        }
    }

    /// Expand the visitor model into a mutable implementation
    pub fn expand_mut(&self) -> TokenStream {
        // Store each of the three parts
        let mut trait_functions: Vec<TokenStream> = Vec::new();
        let mut helpers: Vec<TokenStream> = Vec::new();
        let mut visitables: Vec<TokenStream> = Vec::new();

        // Loop over each type given
        for t in self.types.iter() {
            let elem_name = Ident::new(
                &t.inner_type.ident.to_string().to_case(Case::Snake),
                t.inner_type.ident.span(),
            );
            let elem_type = &t.inner_type;
            let fn_name = format_ident!("visit_{}_mut", elem_name);
            let options = Options::new(&t.attrs.options);

            // Get trait function
            if options.no_default {
                trait_functions.push(quote! {
                    fn #fn_name(&mut self, #elem_name: &mut #elem_type);
                })
            } else {
                trait_functions.push(quote! {
                    fn #fn_name(&mut self, #elem_name: &mut #elem_type) {
                        #fn_name(self, #elem_name)
                    }
                })
            };

            // Get helper function
            if options.has_helper {
                if let Some(inner) = options.helper_tmpl {
                    helpers.push(quote! {
                        pub fn #fn_name<V>(visitor: &mut V, #elem_name: &mut #elem_type)
                        where
                            V: VisitorMut + ?Sized,
                        {
                            #inner
                        }
                    });
                } else {
                    let unused_elem_name = format_ident!("_{}", elem_name);
                    helpers.push(quote! {
                        pub fn #fn_name<V>(_visitor: &mut V, #unused_elem_name: &mut #elem_type)
                        where
                            V: VisitorMut + ?Sized,
                        {
                        }
                    });
                }
            };

            // Make visitable
            visitables.push(quote! {
                impl VisitableMut for #elem_type {
                    fn apply(&mut self, visitor: &mut impl VisitorMut) {
                        visitor.#fn_name(self);
                    }
                }
            });
        }

        // Built complete visitor implementation
        quote! {
            pub trait VisitorMut {
                #(#trait_functions)*
            }

            #(#helpers)*

            trait VisitableMut {
                fn apply(&mut self, visitor: &mut impl VisitorMut);
            }
            #(#visitables)*
        }
    }
}

/// Private struct for dissecting each option passed to a visitor type
struct Options {
    no_default: bool,
    has_helper: bool,
    helper_tmpl: Option<TokenStream>,
}

impl Options {
    fn new(options: &Punctuated<KeyValue, Token![,]>) -> Self {
        // Defaults
        let mut no_default = false;
        let mut has_helper = true;
        let mut helper_tmpl = None;

        // Loop over each option given
        for option in options.iter() {
            // "no_default" turns no_default on
            if option.key == Ident::new("no_default", Span::call_site()) {
                no_default = true;
                continue;
            }

            if option.key == Ident::new("helper_tmpl", Span::call_site()) {
                match &option.value {
                    TokenTree::Ident(ident) if ident == &Ident::new("false", Span::call_site()) => {
                        // "helper_tmpl = false" turns helper template off
                        has_helper = false;
                    }
                    TokenTree::Group(group) => {
                        // Custom helper template was given
                        helper_tmpl = Some(group.stream());
                    }
                    _ => continue,
                }
            }
        }

        Options {
            no_default,
            has_helper,
            helper_tmpl,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use despatma_test_helpers::reformat;
    use pretty_assertions::assert_eq;
    use syn::{parse_quote, parse_str};

    type Result = std::result::Result<(), Box<dyn std::error::Error>>;

    #[test]
    fn parse() {
        let actual: VisitorFunction = parse_quote! {
            #[no_default]
            dyn Button
        };

        let mut expected = VisitorFunction {
            types: Punctuated::new(),
        };

        expected.types.push(parse_quote! {#[no_default] dyn Button});

        assert_eq!(actual, expected);
    }

    #[test]
    fn parse_just_types() -> Result {
        let actual: VisitorFunction = parse_str("Button, dyn Text, Window")?;

        let mut expected = VisitorFunction {
            types: Punctuated::new(),
        };

        expected.types.push(parse_str("Button")?);
        expected.types.push(parse_str("dyn Text")?);
        expected.types.push(parse_str("Window")?);

        assert_eq!(actual, expected);

        Ok(())
    }

    #[test]
    fn parse_mixed() -> Result {
        let actual: VisitorFunction = parse_quote! {
            Button,

            #[tmpl = {trait T {};}]
            Text,

            dyn Window
        };

        let mut expected = VisitorFunction {
            types: Punctuated::new(),
        };

        expected.types.push(parse_str("Button")?);
        expected.types.push(parse_quote! {
            #[tmpl = {trait T {};}]
            Text
        });
        expected.types.push(parse_str("dyn Window")?);

        assert_eq!(actual, expected);

        Ok(())
    }

    #[test]
    fn expand() -> Result {
        let mut input = VisitorFunction {
            types: Punctuated::new(),
        };

        input.types.push(parse_quote! {
            #[helper_tmpl = false]
            Button
        });
        input.types.push(parse_quote! {
            #[no_default]
            dyn Text
        });
        input.types.push(parse_quote! {
           #[helper_tmpl = {
               visitor.visit_button(&window.button);
           }]
           Window
        });

        let actual = input.expand();
        let expected = quote! {
            pub trait Visitor{
                fn visit_button(&mut self, button: &Button) {
                    visit_button(self, button)
                }
                fn visit_text(&mut self, text: &dyn Text);
                fn visit_window(&mut self, window: &Window) {
                    visit_window(self, window)
                }
            }

            pub fn visit_text<V>(_visitor: &mut V, _text: &dyn Text)
            where
                V: Visitor + ?Sized,
            {
            }

            pub fn visit_window<V>(visitor: &mut V, window: &Window)
            where
                V: Visitor + ?Sized,
            {
               visitor.visit_button(&window.button);
            }

            trait Visitable {
                fn apply(&self, visitor: &mut impl Visitor);
            }
            impl Visitable for Button {
                fn apply(&self, visitor: &mut impl Visitor) {
                    visitor.visit_button(self);
                }
            }
            impl Visitable for dyn Text {
                fn apply(&self, visitor: &mut impl Visitor) {
                    visitor.visit_text(self);
                }
            }
            impl Visitable for Window {
                fn apply(&self, visitor: &mut impl Visitor) {
                    visitor.visit_window(self);
                }
            }
        };

        assert_eq!(
            reformat(&actual).lines().collect::<Vec<_>>(),
            reformat(&expected).lines().collect::<Vec<_>>()
        );

        Ok(())
    }
}
