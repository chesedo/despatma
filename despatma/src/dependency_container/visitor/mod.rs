use super::{ChildDependency, Container, Dependency};

mod async_visitor;
mod impl_trait_but_registered_concrete;
mod wiring_visitor;

pub use async_visitor::AsyncVisitor;
pub use impl_trait_but_registered_concrete::ImplTraitButRegisteredConcrete;
pub use wiring_visitor::WiringVisitor;

/// A visitor used to validate the struct that will be turned into a dependency container.
/// If the visitor found any errors then they should be emit in [emit_errors].
pub trait Visit {
    /// Visit the top level container which will be turned into a struct
    fn visit_container(&mut self, container: &Container) {
        visit_container(self, container);
    }

    /// Visit a dependency that was registered
    fn visit_dependency(&mut self, dependency: &Dependency) {
        visit_dependency(self, dependency);
    }

    /// Visit a dependency requested by a registered dependency
    fn visit_child_dependency(&mut self, child_dependency: &ChildDependency) {
        visit_child_dependency(self, child_dependency);
    }

    /// Emit any errors that were found during the visit
    fn emit_errors(self);
}

fn visit_container<V: Visit + ?Sized>(visitor: &mut V, container: &Container) {
    for dependency in container.dependencies.values() {
        visitor.visit_dependency(&dependency.borrow());
    }
}

fn visit_dependency<V: Visit + ?Sized>(visitor: &mut V, dependency: &Dependency) {
    for child_dependency in &dependency.dependencies {
        visitor.visit_child_dependency(child_dependency);
    }
}

fn visit_child_dependency<V: Visit + ?Sized>(
    _visitor: &mut V,
    _child_dependency: &ChildDependency,
) {
}

/// A mutable visitor used to update any dependencies or their children
pub trait VisitMut {
    fn visit_container_mut(&mut self, container: &mut Container) {
        visit_container_mut(self, container);
    }

    fn visit_dependency_mut(&mut self, dependency: &mut Dependency) {
        visit_dependency_mut(self, dependency);
    }

    fn visit_child_dependency_mut(&mut self, child_dependency: &mut ChildDependency) {
        visit_child_dependency_mut(self, child_dependency);
    }
}

fn visit_container_mut<V: VisitMut + ?Sized>(visitor: &mut V, container: &mut Container) {
    for dependency in container.dependencies.values_mut() {
        visitor.visit_dependency_mut(&mut dependency.borrow_mut());
    }
}

fn visit_dependency_mut<V: VisitMut + ?Sized>(visitor: &mut V, dependency: &mut Dependency) {
    for child_dependency in &mut dependency.dependencies {
        visitor.visit_child_dependency_mut(child_dependency);
    }
}

fn visit_child_dependency_mut<V: VisitMut + ?Sized>(
    _visitor: &mut V,
    _child_dependency: &mut ChildDependency,
) {
}
