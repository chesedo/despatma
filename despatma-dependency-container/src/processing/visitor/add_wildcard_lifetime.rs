use syn::{parse_quote, GenericArgument, PathArguments, Type, TypeImplTrait, TypeTraitObject};

use crate::processing::Dependency;

use super::{ErrorVisitorMut, VisitorMut};

/// Add the wildcard lifetime to any return types that might need it.
/// This is for dependencies which requests and returns an impl Trait dependency which has a managed lifetime.
///
/// Needs to be called after lifetimes are extracted.
/// And after dependencies are linked.
/// Needs to be called before boxes are wrapped again.
pub struct AddWildcardLifetime;

impl VisitorMut for AddWildcardLifetime {
    fn visit_dependency_mut(&mut self, dependency: &mut Dependency) {
        let Type::Path(path) = &mut dependency.ty else {
            return;
        };

        let Some(last_segment) = path.path.segments.last_mut() else {
            return;
        };

        let PathArguments::AngleBracketed(generics) = &mut last_segment.arguments else {
            return;
        };

        let deps_needing_wildcard_lifetime: Vec<_> = dependency
            .dependencies
            .iter()
            .filter(|dep| dep.inner.borrow().lifetime.is_managed())
            .map(|dep| dep.inner.borrow().ty.clone())
            .filter_map(|ty| match ty {
                Type::TraitObject(TypeTraitObject { bounds, .. }) => Some(bounds),
                Type::ImplTrait(TypeImplTrait { bounds, .. }) => Some(bounds),
                _ => None,
            })
            .collect();

        for arg in generics.args.iter_mut() {
            let GenericArgument::Type(Type::ImplTrait(type_impl_trait)) = arg else {
                continue;
            };

            if deps_needing_wildcard_lifetime.contains(&type_impl_trait.bounds) {
                type_impl_trait.bounds.push(parse_quote!('_));
            }
        }
    }
}

impl ErrorVisitorMut for AddWildcardLifetime {
    fn new() -> Self {
        Self
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;
    use syn::parse_quote;

    use crate::{
        input,
        processing::{
            self,
            visitor::{ExtractBoxType, ExtractLifetime, LinkDependencies, VisitableMut},
        },
    };

    use super::*;

    #[test]
    fn add_wildcard_lifetime() {
        let mut container: processing::Container = input::Container::from_item_impl(parse_quote!(
            impl Container {
                #[Singleton]
                fn dal(&self) -> Box<dyn DAL> {
                    Box::new(Postgres)
                }

                #[Scoped]
                fn config(&self) -> impl Config {
                    ConfigStruct::parse()
                }

                fn datetime(&self) -> Box<Utc> {
                    Box::new(Utc::now())
                }

                fn service(&self, dal: impl DAL, config: impl Config, datetime: Utc) -> Service<impl DAL, impl Config> {
                    Service::new(dal, config, datetime)
                }
            }
        ))
        .into();

        container.apply_mut(&mut ExtractBoxType);
        container.apply_mut(&mut ExtractLifetime);
        container.apply_mut(&mut LinkDependencies::new());

        assert_eq!(container.dependencies[0].borrow().ty, parse_quote!(dyn DAL));
        assert_eq!(
            container.dependencies[1].borrow().ty,
            parse_quote!(impl Config)
        );
        assert_eq!(container.dependencies[2].borrow().ty, parse_quote!(Utc));
        assert_eq!(
            container.dependencies[3].borrow().ty,
            parse_quote!(Service<impl DAL, impl Config>),
        );

        container.apply_mut(&mut AddWildcardLifetime);

        assert_eq!(container.dependencies[0].borrow().ty, parse_quote!(dyn DAL));
        assert_eq!(
            container.dependencies[1].borrow().ty,
            parse_quote!(impl Config)
        );
        assert_eq!(container.dependencies[2].borrow().ty, parse_quote!(Utc));
        assert_eq!(
            container.dependencies[3].borrow().ty,
            parse_quote!(Service<impl DAL + '_, impl Config + '_>),
        );
    }
}
