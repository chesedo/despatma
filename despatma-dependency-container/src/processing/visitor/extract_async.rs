use crate::processing::Dependency;

use super::{ErrorVisitorMut, VisitorMut};

/// Fix up any dependencies that might be async. It checks whether the registered function is async. Or whether any
/// child dependencies are async. Because if any child dependencies are async, then the parent dependency must also be
/// async. Which this visitor updates correcly.
///
/// Requires child dependencies to be linked first.
pub struct ExtractAsync;

impl VisitorMut for ExtractAsync {
    fn visit_dependency_mut(&mut self, dependency: &mut Dependency) {
        dependency.is_async = dependency.sig.asyncness.is_some();

        for child in dependency.dependencies.iter_mut() {
            let child = &child.inner;
            self.visit_dependency_mut(&mut child.borrow_mut());

            dependency.is_async |= child.borrow().is_async;
        }
    }
}

impl ErrorVisitorMut for ExtractAsync {
    fn new() -> Self {
        Self
    }
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use crate::{
        input,
        processing::{
            self,
            visitor::{ErrorVisitorMut, LinkDependencies, VisitableMut},
        },
    };

    use super::*;

    #[test]
    fn extract_async() {
        let mut container: processing::Container = input::Container::from_item_impl(parse_quote!(
            impl Container {
                // Have this first to make sure child dependencies are considered correctly
                fn service(&self, middle_foo: u32) -> u32 {
                    async_foo
                }

                fn foo(&self) -> u32 {
                    42
                }

                async fn async_foo(&self) -> u32 {
                    42
                }

                fn middle_foo(&self, async_foo: u32) -> u32 {
                    async_foo
                }
            }
        ))
        .into();

        // Test needs them to be linked
        container.apply_mut(&mut LinkDependencies::new());

        assert!(!container.dependencies[0].borrow().is_async);
        assert!(!container.dependencies[1].borrow().is_async);
        assert!(!container.dependencies[2].borrow().is_async);
        assert!(!container.dependencies[3].borrow().is_async);

        container.apply_mut(&mut ExtractAsync);

        assert!(container.dependencies[0].borrow().is_async);
        assert!(!container.dependencies[1].borrow().is_async);
        assert!(container.dependencies[2].borrow().is_async);
        assert!(container.dependencies[3].borrow().is_async);
    }
}
