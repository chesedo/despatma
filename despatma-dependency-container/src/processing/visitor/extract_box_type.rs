use syn::{GenericArgument, PathArguments, Type};

use crate::processing::Dependency;

use super::{ErrorVisitorMut, VisitorMut};

/// Extracts the internal type of any boxed dependency
/// This is needed to fix the Rust lifetimes on any singleton and scoped dependencies
pub struct ExtractBoxType;

impl VisitorMut for ExtractBoxType {
    fn visit_dependency_mut(&mut self, dependency: &mut Dependency) {
        let Type::Path(path) = &dependency.ty else {
            return;
        };

        let Some(last_segment) = path.path.segments.last() else {
            return;
        };

        if last_segment.ident != "Box" {
            return;
        }

        let PathArguments::AngleBracketed(generics) = &last_segment.arguments else {
            return;
        };

        // A box can only have one generic type
        if generics.args.len() != 1 {
            return;
        }

        let GenericArgument::Type(ty) = &generics.args[0] else {
            return;
        };

        dependency.is_boxed = true;
        dependency.ty = ty.clone();
    }
}

impl ErrorVisitorMut for ExtractBoxType {
    fn new() -> Self {
        Self
    }
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use crate::{
        input,
        processing::{self, visitor::VisitableMut},
    };

    use super::*;

    #[test]
    fn extract_box_type() {
        let mut container: processing::Container = input::Container::from_item_impl(parse_quote!(
            impl Container {
                #[Singleton]
                fn singleton(&self) -> Box<dyn DAL> {
                    Box::new(Postgres)
                }

                fn dal(&self) -> std::boxed::Box<dyn DAL> {
                    Box::new(Sqlite)
                }
            }
        ))
        .into();

        assert!(!container.dependencies[0].borrow().is_boxed);
        assert_eq!(
            container.dependencies[0].borrow().ty,
            parse_quote!(Box<dyn DAL>),
        );
        assert!(!container.dependencies[1].borrow().is_boxed);
        assert_eq!(
            container.dependencies[1].borrow().ty,
            parse_quote!(std::boxed::Box<dyn DAL>),
        );

        container.apply_mut(&mut ExtractBoxType);

        assert!(container.dependencies[0].borrow().is_boxed);
        assert_eq!(container.dependencies[0].borrow().ty, parse_quote!(dyn DAL));
        assert!(container.dependencies[1].borrow().is_boxed);
        assert_eq!(container.dependencies[1].borrow().ty, parse_quote!(dyn DAL));
    }
}
