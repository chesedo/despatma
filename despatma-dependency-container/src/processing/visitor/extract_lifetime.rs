use syn::Meta;

use crate::processing::{Dependency, Lifetime};

use super::{ErrorVisitorMut, VisitorMut};

/// Get the lifetime of a dependency from the function attributes
pub struct ExtractLifetime;

impl VisitorMut for ExtractLifetime {
    fn visit_dependency_mut(&mut self, dependency: &mut Dependency) {
        // Remove all lifetime attributes
        dependency.attrs.retain(|attr| {
            let Meta::Path(ref path) = attr.meta else {
                return true;
            };

            if path.segments.len() != 1 {
                return true;
            }

                match path.segments[0].ident.to_string().as_str() {
                    "Transient" => {
                        dependency.lifetime = Lifetime::Transient;
                        false
                    }
                    "Scoped" => {
                        dependency.lifetime = Lifetime::Scoped;
                        false
                    }

                    "Singleton" => {
                        dependency.lifetime = Lifetime::Singleton;
                        false
                    }
                    _ => true,
                }
        });
    }
}

impl ErrorVisitorMut for ExtractLifetime {
    fn new() -> Self {
        Self
    }
    }

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use crate::{
        input,
        processing::{self, visitor::VisitableMut},
    };

    use super::*;

    #[test]
    fn extract_lifetime() {
        let mut container: processing::Container = input::Container::from_item_impl(parse_quote!(
            impl Container {
                #[Singleton]
                fn singleton(&self) -> Singleton {
                    Singleton
                }

                #[Scoped]
                fn scoped(&self) -> Scoped {
                    Scoped
                }

                #[Transient]
                fn transient(&self) -> Transient {
                    Transient
                }

                fn default(&self) -> Default {
                    Default
                }
            }
        ))
        .into();

        assert_eq!(container.dependencies[0].borrow().attrs.len(), 1);
        assert_eq!(
            container.dependencies[0].borrow().lifetime,
            Lifetime::Transient
        );
        assert_eq!(container.dependencies[1].borrow().attrs.len(), 1);
        assert_eq!(
            container.dependencies[1].borrow().lifetime,
            Lifetime::Transient
        );
        assert_eq!(container.dependencies[2].borrow().attrs.len(), 1);
        assert_eq!(
            container.dependencies[2].borrow().lifetime,
            Lifetime::Transient
        );
        assert_eq!(container.dependencies[3].borrow().attrs.len(), 0);
        assert_eq!(
            container.dependencies[3].borrow().lifetime,
            Lifetime::Transient
        );

        container.apply_mut(&mut ExtractLifetime);

        assert_eq!(container.dependencies[0].borrow().attrs.len(), 0);
        assert_eq!(
            container.dependencies[0].borrow().lifetime,
            Lifetime::Singleton
        );
        assert_eq!(container.dependencies[1].borrow().attrs.len(), 0);
        assert_eq!(
            container.dependencies[1].borrow().lifetime,
            Lifetime::Scoped
        );
        assert_eq!(container.dependencies[2].borrow().attrs.len(), 0);
        assert_eq!(
            container.dependencies[2].borrow().lifetime,
            Lifetime::Transient
        );
        assert_eq!(container.dependencies[3].borrow().attrs.len(), 0);
        assert_eq!(
            container.dependencies[3].borrow().lifetime,
            Lifetime::Transient
        );
    }
}
