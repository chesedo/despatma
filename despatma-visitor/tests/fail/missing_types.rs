/// Test when the types are missing
mod library;

use despatma_visitor::visitor;

visitor!(dyn Circle, Rectangle, dyn Sphere, Arc, dyn Cube);

fn main() {}
