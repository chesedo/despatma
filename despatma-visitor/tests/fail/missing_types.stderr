error[E0405]: cannot find trait `Circle` in this scope
 --> tests/fail/missing_types.rs:6:14
  |
6 | visitor!(dyn Circle, Rectangle, dyn Sphere, Arc, dyn Cube);
  |              ^^^^^^ not found in this scope
  |
help: consider importing this trait
  |
4 + use crate::library::shapes::Circle;
  |

error[E0412]: cannot find type `Rectangle` in this scope
 --> tests/fail/missing_types.rs:6:22
  |
6 | visitor!(dyn Circle, Rectangle, dyn Sphere, Arc, dyn Cube);
  |                      ^^^^^^^^^ not found in this scope
  |
help: consider importing this struct
  |
4 + use crate::library::shapes::Rectangle;
  |

error[E0405]: cannot find trait `Sphere` in this scope
 --> tests/fail/missing_types.rs:6:37
  |
6 | visitor!(dyn Circle, Rectangle, dyn Sphere, Arc, dyn Cube);
  |                                     ^^^^^^ not found in this scope
  |
help: consider importing this trait
  |
4 + use crate::library::shapes::Sphere;
  |

error[E0412]: cannot find type `Arc` in this scope
 --> tests/fail/missing_types.rs:6:45
  |
6 | visitor!(dyn Circle, Rectangle, dyn Sphere, Arc, dyn Cube);
  |                                             ^^^ not found in this scope
  |
help: consider importing one of these structs
  |
4 + use crate::library::shapes::Arc;
  |
4 + use std::sync::Arc;
  |

error[E0405]: cannot find trait `Cube` in this scope
 --> tests/fail/missing_types.rs:6:54
  |
6 | visitor!(dyn Circle, Rectangle, dyn Sphere, Arc, dyn Cube);
  |                                                      ^^^^ not found in this scope
  |
help: consider importing this trait
  |
4 + use crate::library::shapes::Cube;
  |
